PR: https://github.com/odoo/odoo/pull/219608

From: 01434d997693cf5416c0e34b0194c562e0da22b3
From: Abdelrahman Mahmoud (amah)
Date: 2025-08-14 09:24:18

Breaking data model changes scores: del:12 + add:2, change matches:
+    attendance_ids_1st_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 1st Week',
+    attendance_ids_2nd_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 2nd Week',
-    two_weeks_calendar = fields.Boolean(string="Calendar in 2 weeks mode")
-    two_weeks_explanation = fields.Char('Explanation', compute="_compute_two_weeks_explanation")
-    flexible_hours = fields.Boolean(string="Flexible Hours",
-    hours_per_week = fields.Float(
-    full_time_required_hours = fields.Float(
-    is_fulltime = fields.Boolean(compute='_compute_work_time_rate', string="Is Full Time")
-    work_resources_count = fields.Integer("Work Resources count", compute='_compute_work_resources_count')
-    schedule_type = fields.Selection(
-    attendance_ids_1st_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 1st Week',
-    attendance_ids_2nd_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 2nd Week',

Total Changes: 515

[REF] resource: reorganise resource.calandar file

This commit just moves methods to organise them better before starting the
proper refactoring. This just allows to avoid noise for the real refactoring
review.

* merge _check_attendance_ids constrains in resource.calendar

* simplify _get_days_pet_week method.

* remove deprecated private methods

* unify hours_per_days and per week computation

Base the computation on the global attendance only.
Compute per_week and base the per_hour on per_week.
Note that per_hour cannot depend directly form the per_week because of
rounding issues.

* remove _get_max_number_of_hours as only used once in localisation

Task-4951007

closes odoo/odoo#219608

Related: odoo/enterprise#90494
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/resource/models/resource_calendar.py
+++ b/addons/resource/models/resource_calendar.py
@@ -10,7 +10,7 @@ from dateutil.relativedelta import relativedelta
 from dateutil.rrule import DAILY, rrule
 from pytz import timezone, utc
 
-from odoo import _, api, fields, models
+from odoo import api, fields, models
 from odoo.exceptions import ValidationError
 from odoo.fields import Command, Domain
 from odoo.tools import date_utils, float_compare, ormcache
@@ -24,10 +24,10 @@ from odoo.addons.base.models.res_partner import _tz_get
 class ResourceCalendar(models.Model):
     """ Calendar model for a resource. It has
 
-     - attendance_ids: list of resource.calendar.attendance that are a working
-                       interval in a given weekday.
-     - leave_ids: list of leaves linked to this calendar. A leave can be general
-                  or linked to a specific resource, depending on its resource_id.
+    - attendance_ids: list of resource.calendar.attendance that are a working
+                    interval in a given weekday.
+    - leave_ids: list of leaves linked to this calendar. A leave can be general
+                or linked to a specific resource, depending on its resource_id.
 
     All methods in this class use intervals. An interval is a tuple holding
     (begin_datetime, end_datetime). A list of intervals is therefore a list of
@@ -39,11 +39,12 @@ class ResourceCalendar(models.Model):
     def default_get(self, fields):
         res = super().default_get(fields)
         if not res.get('name') and res.get('company_id'):
-            res['name'] = _('Working Hours of %s', self.env['res.company'].browse(res['company_id']).name)
+            res['name'] = self.env._('Working Hours of %s', self.env['res.company'].browse(res['company_id']).name)
         if 'attendance_ids' in fields and not res.get('attendance_ids'):
             company_id = res.get('company_id', self.env.company.id)
             company = self.env['res.company'].browse(company_id)
             res["attendance_ids"] = self._get_default_attendance_ids(company)
+            res["two_weeks_calendar"] = company.resource_calendar_id.two_weeks_calendar
         if 'full_time_required_hours' in fields and not res.get('full_time_required_hours'):
             company_id = res.get('company_id', self.env.company.id)
             company = self.env['res.company'].browse(company_id)
@@ -53,59 +54,83 @@ class ResourceCalendar(models.Model):
     name = fields.Char(required=True)
     active = fields.Boolean("Active", default=True,
                             help="If the active field is set to false, it will allow you to hide the Working Time without removing it.")
-    company_id = fields.Many2one(
-        'res.company', 'Company', domain=lambda self: [('id', 'in', self.env.companies.ids)],
-        default=lambda self: self.env.company, index='btree_not_null')
     attendance_ids = fields.One2many(
         'resource.calendar.attendance', 'calendar_id', 'Working Time',
         compute='_compute_attendance_ids', store=True, readonly=False, copy=True)
+    attendance_ids_1st_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 1st Week',
+        compute="_compute_two_weeks_attendance", inverse="_inverse_two_weeks_calendar")
+    attendance_ids_2nd_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 2nd Week',
+        compute="_compute_two_weeks_attendance", inverse="_inverse_two_weeks_calendar")
+    company_id = fields.Many2one(
+        'res.company', 'Company', domain=lambda self: [('id', 'in', self.env.companies.ids)],
+        default=lambda self: self.env.company, index='btree_not_null')
     leave_ids = fields.One2many(
         'resource.calendar.leaves', 'calendar_id', 'Time Off')
+    schedule_type = fields.Selection(
+        [
+            ('flexible', 'Flexible'),
+            ('fully_fixed', 'Fully Fixed'),
+        ],
+        string='Schedule Type',
+        required=True,
+        default='fully_fixed',
+        help="Choose which level of definition you want to define on your Schedule\n"
+            "- Flexible : Define an amount of hours to work on the week.\n"
+            "- Fully Fixed : define the days, periods and the start & end time for each period of the day",
+    )
+    flexible_hours = fields.Boolean(string="Flexible Hours",
+        compute="_compute_flexible_hours", inverse="_inverse_flexible_hours", store=True,
+        help="When enabled, it will allow employees to work flexibly, without relying on the company's working schedule (working hours).")
+    full_time_required_hours = fields.Float(
+        string="Company Full Time",
+        compute="_compute_full_time_required_hours", store=True, readonly=False,
+        help="Number of hours to work on the company schedule to be considered as fulltime.")
     global_leave_ids = fields.One2many(
         'resource.calendar.leaves', 'calendar_id', 'Global Time Off',
         compute='_compute_global_leave_ids', store=True, readonly=False,
         domain=[('resource_id', '=', False)], copy=True,
     )
     hours_per_day = fields.Float("Average Hour per Day", store=True, compute="_compute_hours_per_day", digits=(2, 2), readonly=False,
-                                 help="Average hours per day a resource is supposed to work with this calendar.")
+        help="Average hours per day a resource is supposed to work with this calendar.")
+    hours_per_week = fields.Float(
+        string="Hours per Week",
+        compute="_compute_hours_per_week", store=True, readonly=False, copy=False)
+    is_fulltime = fields.Boolean(compute='_compute_work_time_rate', string="Is Full Time")
+    two_weeks_calendar = fields.Boolean(string="Calendar in 2 weeks mode")
+    two_weeks_explanation = fields.Char('Explanation', compute="_compute_two_weeks_explanation")
     tz = fields.Selection(
         _tz_get, string='Timezone', required=True,
         default=lambda self: self.env.context.get('tz') or self.env.user.tz or self.env.ref('base.user_admin').tz or 'UTC',
         help="This field is used in order to define in which timezone the resources will work.")
     tz_offset = fields.Char(compute='_compute_tz_offset', string='Timezone offset')
-    two_weeks_calendar = fields.Boolean(string="Calendar in 2 weeks mode")
-    two_weeks_explanation = fields.Char('Explanation', compute="_compute_two_weeks_explanation")
-    flexible_hours = fields.Boolean(string="Flexible Hours",
-        compute="_compute_flexible_hours", inverse="_inverse_flexible_hours",
-        help="When enabled, it will allow employees to work flexibly, without relying on the company's working schedule (working hours).")
-    hours_per_week = fields.Float(
-        string="Hours per Week",
-        compute="_compute_hours_per_week", store=True, readonly=False)
-    full_time_required_hours = fields.Float(
-        string="Company Full Time",
-        compute="_compute_full_time_required_hours", store=True, readonly=False,
-        help="Number of hours to work on the company schedule to be considered as fulltime.")
-    is_fulltime = fields.Boolean(compute='_compute_work_time_rate', string="Is Full Time")
+    work_resources_count = fields.Integer("Work Resources count", compute='_compute_work_resources_count')
     work_time_rate = fields.Float(string='Work Time Rate', compute='_compute_work_time_rate', search='_search_work_time_rate',
         help='Work time rate versus full time working schedule, should be between 0 and 100 %.')
-    work_resources_count = fields.Integer("Work Resources count", compute='_compute_work_resources_count')
-    schedule_type = fields.Selection(
-        [
-            ('flexible', 'Flexible'),
-            ('fully_fixed', 'Fully Fixed'),
-        ],
-        string='Schedule Type',
-        required=True,
-        default='fully_fixed',
-        help="Choose which level of definition you want to define on your Schedule\n"
-            "- Flexible : Define an amount of hours to work on the week.\n"
-            "- Fully Fixed : define the days, periods and the start & end time for each period of the day",
-    )
-    attendance_ids_1st_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 1st Week',
-        compute="_compute_two_weeks_attendance", inverse="_inverse_two_weeks_calendar")
 
-    attendance_ids_2nd_week = fields.One2many('resource.calendar.attendance', 'calendar_id', 'Working Time 2nd Week',
-        compute="_compute_two_weeks_attendance", inverse="_inverse_two_weeks_calendar")
+    # --------------------------------------------------
+    # Constrains
+    # --------------------------------------------------
+
+    @api.constrains('attendance_ids')
+    def _check_attendance_ids(self):
+        for res_calendar in self:
+            if (res_calendar.two_weeks_calendar and
+                    res_calendar.attendance_ids.filtered(lambda a: a.display_type == 'line_section') and
+                    not res_calendar.attendance_ids.sorted('sequence')[0].display_type):
+                raise ValidationError(self.env._("In a calendar with 2 weeks mode, all periods need to be in the sections."))
+
+            # Avoid superimpose in attendance
+            attendance_ids = res_calendar.attendance_ids.filtered(
+                lambda attendance: not attendance.resource_id and attendance.display_type is False)
+            if res_calendar.two_weeks_calendar:
+                res_calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '0'))
+                res_calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '1'))
+            else:
+                res_calendar._check_overlap(attendance_ids)
+
+    # --------------------------------------------------
+    # Compute Methods
+    # --------------------------------------------------
 
     @api.depends('two_weeks_calendar')
     def _compute_two_weeks_attendance(self):
@@ -119,15 +144,12 @@ class ResourceCalendar(models.Model):
         for calendar in self:
             if not calendar.two_weeks_calendar:
                 continue
-            calendar.attendance_ids = [Command.clear()] + calendar.attendance_ids_1st_week + calendar.attendance_ids_2nd_week
+            calendar.attendance_ids = calendar.attendance_ids_1st_week + calendar.attendance_ids_2nd_week
 
     @api.depends('hours_per_week', 'company_id.resource_calendar_id.hours_per_week')
     def _compute_full_time_required_hours(self):
         for calendar in self.filtered("company_id"):
-            if calendar != calendar.company_id.resource_calendar_id:
-                calendar.full_time_required_hours = calendar.company_id.resource_calendar_id.hours_per_week
-            else:
-                calendar.full_time_required_hours = calendar.hours_per_week
+            calendar.full_time_required_hours = calendar.company_id.resource_calendar_id.hours_per_week
 
     @api.depends("schedule_type")
     def _compute_flexible_hours(self):
@@ -138,147 +160,119 @@ class ResourceCalendar(models.Model):
         for calendar in self:
             calendar.schedule_type = 'flexible' if calendar.flexible_hours else 'fully_fixed'
 
-    def _compute_work_resources_count(self):
-        count_data = self.env['resource.resource']._read_group(
-            domain=[('calendar_id', 'in', self.ids)],
-            groupby=['calendar_id'],
-            aggregates=['__count'])
-        mapped_counts = {resource_calendar.id: count for resource_calendar, count in count_data}
-        for calendar in self:
-            calendar.work_resources_count = mapped_counts.get(calendar.id, 0)
-
-    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to', 'two_weeks_calendar', 'flexible_hours')
-    def _compute_hours_per_day(self):
-        for calendar in self:
-            if calendar.flexible_hours:
-                continue
-            attendances = calendar._get_global_attendances()
-            calendar.hours_per_day = calendar._get_hours_per_day(attendances)
-
-    @api.depends('attendance_ids.hour_from', 'attendance_ids.hour_to')
-    def _compute_hours_per_week(self):
-        for calendar in self:
-            sum_hours = sum(
-                (attendance.hour_to - attendance.hour_from) for attendance in calendar.attendance_ids if attendance._is_work_period())
-            calendar.hours_per_week = sum_hours / 2 if calendar.two_weeks_calendar else sum_hours
-
-    @api.depends('hours_per_week', 'full_time_required_hours')
-    def _compute_work_time_rate(self):
-        for calendar in self:
-            if calendar.full_time_required_hours:
-                calendar.work_time_rate = calendar.hours_per_week / calendar.full_time_required_hours * 100
-            else:
-                calendar.work_time_rate = 100
-
-            calendar.is_fulltime = float_compare(calendar.full_time_required_hours, calendar.hours_per_week, 3) == 0
-
-    @api.model
-    def _search_work_time_rate(self, operator, value):
-        if operator in ('in', 'not in'):
-            if not all(isinstance(v, int) for v in value):
-                return NotImplemented
-        elif operator in ('<', '>'):
-            if not isinstance(value, int):
-                return NotImplemented
-        else:
-            return NotImplemented
-
-        calendar_ids = self.env['resource.calendar'].search([])
-        if operator == 'in':
-            calender = calendar_ids.filtered(lambda m: m.work_time_rate in value)
-        elif operator == 'not in':
-            calender = calendar_ids.filtered(lambda m: m.work_time_rate not in value)
-        elif operator == '<':
-            calender = calendar_ids.filtered(lambda m: m.work_time_rate < value)
-        elif operator == '>':
-            calender = calendar_ids.filtered(lambda m: m.work_time_rate > value)
-        return [('id', 'in', calender.ids)]
-
     @api.depends('company_id')
     def _compute_attendance_ids(self):
-        for calendar in self.filtered(lambda c: not c._origin or c._origin.company_id != c.company_id and c.company_id):
+        for calendar in self.filtered(lambda c: not c._origin or (c._origin.company_id != c.company_id and c.company_id)):
             company_calendar = calendar.company_id.resource_calendar_id
             calendar.update({
                 'two_weeks_calendar': company_calendar.two_weeks_calendar,
                 'tz': company_calendar.tz,
                 'attendance_ids': [(5, 0, 0)] + [
-                    (0, 0, attendance._copy_attendance_vals()) for attendance in company_calendar.attendance_ids if not attendance.resource_id]
+                    (0, 0, attendance._copy_attendance_vals()) for attendance in company_calendar.attendance_ids if not attendance.resource_id],
             })
 
+    @api.onchange('attendance_ids')
+    def _onchange_attendance_ids(self):
+        if not self.two_weeks_calendar:
+            return
+
+        even_week_seq = self.attendance_ids.filtered(lambda att: att.display_type == 'line_section' and att.week_type == '0')
+        odd_week_seq = self.attendance_ids.filtered(lambda att: att.display_type == 'line_section' and att.week_type == '1')
+        if len(even_week_seq) != 1 or len(odd_week_seq) != 1:
+            raise ValidationError(self.env._("You can't delete section between weeks."))
+
+        even_week_seq = even_week_seq.sequence
+        odd_week_seq = odd_week_seq.sequence
+
+        for line in self.attendance_ids.filtered(lambda att: att.display_type is False):
+            if even_week_seq > odd_week_seq:
+                line.week_type = '1' if even_week_seq > line.sequence else '0'
+            else:
+                line.week_type = '0' if odd_week_seq > line.sequence else '1'
+
     @api.depends('company_id')
     def _compute_global_leave_ids(self):
         for calendar in self.filtered(lambda c: not c._origin or c._origin.company_id != c.company_id):
             calendar.update({
                 'global_leave_ids': [(5, 0, 0)] + [
-                    (0, 0, leave._copy_leave_vals()) for leave in calendar.company_id.resource_calendar_id.global_leave_ids]
+                    (0, 0, leave._copy_leave_vals()) for leave in calendar.company_id.resource_calendar_id.global_leave_ids],
             })
 
-    @api.depends('tz')
-    def _compute_tz_offset(self):
-        for calendar in self:
-            calendar.tz_offset = datetime.now(timezone(calendar.tz or 'GMT')).strftime('%z')
-
-    def copy_data(self, default=None):
-        vals_list = super().copy_data(default=default)
-        return [dict(vals, name=self.env._("%s (copy)", calendar.name)) for calendar, vals in zip(self, vals_list)]
+    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to', 'two_weeks_calendar', 'flexible_hours')
+    def _compute_hours_per_day(self):
+        """ Compute the average hours per day.
+            Cannot directly depend on hours_per_week because of rounding issues. """
+        for calendar in self.filtered(lambda c: not c.flexible_hours):
+            calendar.hours_per_day = float_round(calendar._get_hours_per_day(), precision_digits=2)
 
-    @api.constrains('attendance_ids')
-    def _check_attendance_ids(self):
-        for resource in self:
-            if (resource.two_weeks_calendar and
-                    resource.attendance_ids.filtered(lambda a: a.display_type == 'line_section') and
-                    not resource.attendance_ids.sorted('sequence')[0].display_type):
-                raise ValidationError(_("In a calendar with 2 weeks mode, all periods need to be in the sections."))
+    @api.depends('attendance_ids', 'attendance_ids.hour_from', 'attendance_ids.hour_to', 'two_weeks_calendar', 'flexible_hours')
+    def _compute_hours_per_week(self):
+        """ Compute the average hours per week """
+        for calendar in self.filtered(lambda c: not c.flexible_hours):
+            calendar.hours_per_week = float_round(calendar._get_hours_per_week(), precision_digits=2)
 
     @api.depends('two_weeks_calendar')
     def _compute_two_weeks_explanation(self):
         today = fields.Date.today()
         week_type = self.env['resource.calendar.attendance'].get_week_type(today)
-        week_type_str = _("even") if week_type else _("odd")
+        week_type_str = self.env._("even") if week_type else self.env._("odd")
         first_day = date_utils.start_of(today, 'week')
         last_day = date_utils.end_of(today, 'week')
-        self.two_weeks_explanation = _(
+        self.two_weeks_explanation = self.env._(
             "The current week (from %(first_day)s to %(last_day)s) corresponds to %(number)s week.",
             first_day=first_day,
             last_day=last_day,
             number=week_type_str,
         )
 
-    def _get_global_attendances(self):
-        return self.attendance_ids.filtered(lambda attendance:
-            attendance.day_period != 'lunch'
-            and not attendance.date_from and not attendance.date_to
-            and not attendance.resource_id and not attendance.display_type)
+    @api.depends('tz')
+    def _compute_tz_offset(self):
+        for calendar in self:
+            calendar.tz_offset = datetime.now(timezone(calendar.tz or 'GMT')).strftime('%z')
 
-    def _get_hours_per_day(self, attendances):
-        """
-        Calculate the average hours worked per workday.
-        """
-        if not attendances:
-            return 0
+    def _compute_work_resources_count(self):
+        resources_per_calendar = dict(self.env['resource.resource']._read_group(
+            domain=[('calendar_id', 'in', self.ids)],
+            groupby=['calendar_id'],
+            aggregates=['__count']))
+        for calendar in self:
+            calendar.work_resources_count = resources_per_calendar.get(calendar, 0)
 
-        hour_count = 0.0
-        for attendance in attendances:
-            hour_count += attendance.hour_to - attendance.hour_from
+    @api.depends('hours_per_week', 'full_time_required_hours')
+    def _compute_work_time_rate(self):
+        for calendar in self:
+            if calendar.full_time_required_hours:
+                calendar.work_time_rate = calendar.hours_per_week / calendar.full_time_required_hours * 100
+            else:
+                calendar.work_time_rate = 100
 
-        if self.two_weeks_calendar:
-            number_of_days = len(set(attendances.filtered(lambda cal: cal.week_type == '1').mapped('dayofweek')))
-            number_of_days += len(set(attendances.filtered(lambda cal: cal.week_type == '0').mapped('dayofweek')))
+            calendar.is_fulltime = float_compare(calendar.full_time_required_hours, calendar.hours_per_week, 3) == 0
+
+    @api.model
+    def _search_work_time_rate(self, operator, value):
+        if operator in ('in', 'not in'):
+            if not all(isinstance(v, int) for v in value):
+                return NotImplemented
+        elif operator in ('<', '>'):
+            if not isinstance(value, int):
+                return NotImplemented
         else:
-            number_of_days = len(set(attendances.mapped('dayofweek')))
+            return NotImplemented
 
-        if not number_of_days:
-            return 0
+        calendar_ids = self.env['resource.calendar']._search(Domain('work_time', operator, value))
+        return [('id', 'in', calendar_ids)]
 
-        return float_round(hour_count / float(number_of_days), precision_digits=2)
+    # --------------------------------------------------
+    # Overrides
+    # --------------------------------------------------
 
-    def _get_days_per_week(self):
-        # If the employee didn't work a full day, it is still counted, i.e. 19h / week (M/T/W(half day)) -> 3 days
-        self.ensure_one()
-        days = len(set(self.attendance_ids.filtered(
-            lambda attendance_id: attendance_id._is_work_period() and attendance_id.duration_hours)
-            .mapped(lambda attendance_id: f"{attendance_id.week_type} {attendance_id.dayofweek}")))
-        return days / 2 if self.two_weeks_calendar else days
+    def copy_data(self, default=None):
+        vals_list = super().copy_data(default=default)
+        return [dict(vals, name=self.env._("%s (copy)", calendar.name)) for calendar, vals in zip(self, vals_list)]
+
+    # --------------------------------------------------
+    # Actions
+    # --------------------------------------------------
 
     def switch_calendar_type(self):
         self.ensure_one()
@@ -315,48 +309,6 @@ class ResourceCalendar(models.Model):
             self.attendance_ids.unlink()
             self.attendance_ids = self._get_default_attendance_ids(self.company_id)
 
-    @api.onchange('attendance_ids')
-    def _onchange_attendance_ids(self):
-        if not self.two_weeks_calendar:
-            return
-
-        even_week_seq = self.attendance_ids.filtered(lambda att: att.display_type == 'line_section' and att.week_type == '0')
-        odd_week_seq = self.attendance_ids.filtered(lambda att: att.display_type == 'line_section' and att.week_type == '1')
-        if len(even_week_seq) != 1 or len(odd_week_seq) != 1:
-            raise ValidationError(_("You can't delete section between weeks."))
-
-        even_week_seq = even_week_seq.sequence
-        odd_week_seq = odd_week_seq.sequence
-
-        for line in self.attendance_ids.filtered(lambda att: att.display_type is False):
-            if even_week_seq > odd_week_seq:
-                line.week_type = '1' if even_week_seq > line.sequence else '0'
-            else:
-                line.week_type = '0' if odd_week_seq > line.sequence else '1'
-
-    def _check_overlap(self, attendance_ids):
-        """ attendance_ids correspond to attendance of a week,
-            will check for each day of week that there are no superimpose. """
-        result = []
-        for attendance in attendance_ids.filtered(lambda att: not att.date_from and not att.date_to):
-            # 0.000001 is added to each start hour to avoid to detect two contiguous intervals as superimposing.
-            # Indeed Intervals function will join 2 intervals with the start and stop hour corresponding.
-            result.append((int(attendance.dayofweek) * 24 + attendance.hour_from + 0.000001, int(attendance.dayofweek) * 24 + attendance.hour_to, attendance))
-
-        if len(Intervals(result)) != len(result):
-            raise ValidationError(_("Attendances can't overlap."))
-
-    @api.constrains('attendance_ids')
-    def _check_attendance(self):
-        # Avoid superimpose in attendance
-        for calendar in self:
-            attendance_ids = calendar.attendance_ids.filtered(lambda attendance: not attendance.resource_id and attendance.display_type is False)
-            if calendar.two_weeks_calendar:
-                calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '0'))
-                calendar._check_overlap(attendance_ids.filtered(lambda attendance: attendance.week_type == '1'))
-            else:
-                calendar._check_overlap(attendance_ids)
-
     # --------------------------------------------------
     # Computation API
     # --------------------------------------------------
@@ -431,7 +383,6 @@ class ResourceCalendar(models.Model):
                 else:
                     base_result.append((day_from, day_to, attendance))
 
-
         # Copy the result localized once per necessary timezone
         # Strictly speaking comparing start_dt < time or start_dt.astimezone(tz) < time
         # should always yield the same result. however while working with dates it is easier
@@ -441,7 +392,7 @@ class ResourceCalendar(models.Model):
                 min(bounds_per_tz[tz][1], tz.localize(val[1])),
                 val[2])
                     for val in base_result]
-            for tz in resources_per_tz.keys()
+            for tz in resources_per_tz
         }
         result_per_resource_id = dict()
         for tz, resources in resources_per_tz.items():
@@ -530,7 +481,7 @@ class ResourceCalendar(models.Model):
         if resource is None:
             resource = self.env['resource.resource']
         return self._leave_intervals_batch(
-            start_dt, end_dt, resources=resource, domain=domain, tz=tz
+            start_dt, end_dt, resources=resource, domain=domain, tz=tz,
         )[resource.id]
 
     def _leave_intervals_batch(self, start_dt, end_dt, resources=None, domain=None, tz=None, any_calendar=False):
@@ -558,7 +509,7 @@ class ResourceCalendar(models.Model):
         ]
 
         # retrieve leave intervals in (start_dt, end_dt)
-        result = defaultdict(lambda: [])
+        result = defaultdict(list)
         tz_dates = {}
         all_leaves = self.env['resource.calendar.leaves'].search(domain)
         for leave in all_leaves:
@@ -571,15 +522,15 @@ class ResourceCalendar(models.Model):
                     continue
                 tz = tz if tz else timezone((resource or self).tz)
                 if (tz, start_dt) in tz_dates:
-                    start = tz_dates[(tz, start_dt)]
+                    start = tz_dates[tz, start_dt]
                 else:
                     start = start_dt.astimezone(tz)
-                    tz_dates[(tz, start_dt)] = start
+                    tz_dates[tz, start_dt] = start
                 if (tz, end_dt) in tz_dates:
-                    end = tz_dates[(tz, end_dt)]
+                    end = tz_dates[tz, end_dt]
                 else:
                     end = end_dt.astimezone(tz)
-                    tz_dates[(tz, end_dt)] = end
+                    tz_dates[tz, end_dt] = end
                 dt0 = leave_date_from.astimezone(tz)
                 dt1 = leave_date_to.astimezone(tz)
                 if leave_resource and leave_resource._is_fully_flexible():
@@ -602,16 +553,15 @@ class ResourceCalendar(models.Model):
             return {
                 r.id: (attendance_intervals[r.id] - leave_intervals[r.id]) for r in resources_list
             }
-        else:
-            return {
-                r.id: attendance_intervals[r.id] for r in resources_list
-            }
+        return {
+            r.id: attendance_intervals[r.id] for r in resources_list
+        }
 
     def _unavailable_intervals(self, start_dt, end_dt, resource=None, domain=None, tz=None):
         if resource is None:
             resource = self.env['resource.resource']
         return self._unavailable_intervals_batch(
-            start_dt, end_dt, resources=resource, domain=domain, tz=tz
+            start_dt, end_dt, resources=resource, domain=domain, tz=tz,
         )[resource.id]
 
     def _unavailable_intervals_batch(self, start_dt, end_dt, resources=None, domain=None, tz=None):
@@ -631,7 +581,7 @@ class ResourceCalendar(models.Model):
             # start + flatten(intervals) + end
             work_intervals = [start_dt] + list(chain.from_iterable(work_intervals)) + [end_dt]
             # put it back to UTC
-            work_intervals = list(map(lambda dt: dt.astimezone(utc), work_intervals))
+            work_intervals = [dt.astimezone(utc) for dt in work_intervals]
             # pick groups of two
             work_intervals = list(zip(work_intervals[0::2], work_intervals[1::2]))
             result[resource.id] = work_intervals
@@ -641,6 +591,18 @@ class ResourceCalendar(models.Model):
     # Private Methods / Helpers
     # --------------------------------------------------
 
+    def _check_overlap(self, attendance_ids):
+        """ attendance_ids correspond to attendance of a week,
+            will check for each day of week that there are no superimpose. """
+        result = []
+        for attendance in attendance_ids.filtered(lambda att: not att.date_from and not att.date_to):
+            # 0.000001 is added to each start hour to avoid to detect two contiguous intervals as superimposing.
+            # Indeed Intervals function will join 2 intervals with the start and stop hour corresponding.
+            result.append((int(attendance.dayofweek) * 24 + attendance.hour_from + 0.000001, int(attendance.dayofweek) * 24 + attendance.hour_to, attendance))
+
+        if len(Intervals(result)) != len(result):
+            raise ValidationError(self.env._("Attendances can't overlap."))
+
     def _get_attendance_intervals_days_data(self, attendance_intervals):
         """
         helper function to compute duration of `intervals` that have
@@ -669,49 +631,6 @@ class ResourceCalendar(models.Model):
             'hours': sum(day_hours.values()),
         }
 
-    def _get_days_data(self, intervals, day_total):
-        """
-        helper function to compute duration of `intervals`
-        expressed in days and hours.
-        `day_total` is a dict {date: n_hours} with the number of hours for each day.
-        """
-        day_hours = defaultdict(float)
-        for start, stop, _meta in intervals:
-            day_hours[start.date()] += (stop - start).total_seconds() / 3600
-
-        # compute number of days the hours span over
-        days = float_round(sum(
-            day_hours[day] / day_total[day] if day_total[day] else 0
-            for day in day_hours
-        ), precision_rounding=0.001)
-        return {
-            'days': days,
-            'hours': sum(day_hours.values()),
-        }
-
-    def _get_resources_day_total(self, from_datetime, to_datetime, resources=None):
-        """
-        @return dict with hours of attendance in each day between `from_datetime` and `to_datetime`
-        """
-        self.ensure_one()
-        if not resources:
-            resources = self.env['resource.resource']
-            resources_list = [resources]
-        else:
-            resources_list = list(resources) + [self.env['resource.resource']]
-        # total hours per day:  retrieve attendances with one extra day margin,
-        # in order to compute the total hours on the first and last days
-        from_full = from_datetime - timedelta(days=1)
-        to_full = to_datetime + timedelta(days=1)
-        intervals = self._attendance_intervals_batch(from_full, to_full, resources=resources)
-
-        result = defaultdict(lambda: defaultdict(float))
-        for resource in resources_list:
-            day_total = result[resource.id]
-            for start, stop, _meta in intervals[resource.id]:
-                day_total[start.date()] += (stop - start).total_seconds() / 3600
-        return result
-
     def _get_closest_work_time(self, dt, match_end=False, resource=None, search_range=None, compute_leaves=True):
         """Return the closest work interval boundary within the search range.
         Consider only starts of intervals unless `match_end` is True. It will then only consider
@@ -728,8 +647,8 @@ class ResourceCalendar(models.Model):
         if resource is None:
             resource = self.env['resource.resource']
 
-        if not dt.tzinfo or search_range and not (search_range[0].tzinfo and search_range[1].tzinfo):
-            raise ValueError('Provided datetimes needs to be timezoned')
+        if not dt.tzinfo or (search_range and not (search_range[0].tzinfo and search_range[1].tzinfo)):
+            raise ValueError(self.env._('Provided datetimes needs to be timezoned'))
 
         dt = dt.astimezone(timezone(tz))
 
@@ -747,6 +666,37 @@ class ResourceCalendar(models.Model):
         )
         return interval_dt(work_intervals[0]) if work_intervals else None
 
+    def _get_days_per_week(self):
+        # If the employee didn't work a full day, it is still counted, i.e. 19h / week (M/T/W(half day)) -> 3 days
+        self.ensure_one()
+        attendances = self._get_global_attendances()
+        if self.two_weeks_calendar:
+            number_of_days = len(set(attendances.filtered(lambda cal: cal.week_type == '1').mapped('dayofweek')))
+            number_of_days += len(set(attendances.filtered(lambda cal: cal.week_type == '0').mapped('dayofweek')))
+        else:
+            number_of_days = len(set(attendances.mapped('dayofweek')))
+        return number_of_days / 2 if self.two_weeks_calendar else number_of_days
+
+    def _get_hours_per_week(self):
+        """ Calculate the average hours worked per week. """
+        self.ensure_one()
+        hour_count = 0.0
+        for attendance in self._get_global_attendances():
+            hour_count += attendance.hour_to - attendance.hour_from
+        return hour_count / 2 if self.two_weeks_calendar else hour_count
+
+    def _get_hours_per_day(self):
+        """ Calculate the average hours worked per workday. """
+        hour_per_week = self._get_hours_per_week()
+        number_of_days = self._get_days_per_week()
+        return hour_per_week / number_of_days if number_of_days else 0
+
+    def _get_global_attendances(self):
+        return self.attendance_ids.filtered(lambda attendance:
+            attendance.day_period != 'lunch'
+            and not attendance.date_from and not attendance.date_to
+            and not attendance.resource_id and not attendance.display_type)
+
     def _get_unusual_days(self, start_dt, end_dt, company_id=False):
         if not self:
             return {}
@@ -783,21 +733,21 @@ class ResourceCalendar(models.Model):
                 for attendance in attendances
             ]
         return [
-            Command.create({'name': _('Monday Morning'), 'dayofweek': '0', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': _('Monday Lunch'), 'dayofweek': '0', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': _('Monday Afternoon'), 'dayofweek': '0', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': _('Tuesday Morning'), 'dayofweek': '1', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': _('Tuesday Lunch'), 'dayofweek': '1', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': _('Tuesday Afternoon'), 'dayofweek': '1', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': _('Wednesday Morning'), 'dayofweek': '2', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': _('Wednesday Lunch'), 'dayofweek': '2', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': _('Wednesday Afternoon'), 'dayofweek': '2', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': _('Thursday Morning'), 'dayofweek': '3', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': _('Thursday Lunch'), 'dayofweek': '3', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': _('Thursday Afternoon'), 'dayofweek': '3', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
-            Command.create({'name': _('Friday Morning'), 'dayofweek': '4', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
-            Command.create({'name': _('Friday Lunch'), 'dayofweek': '4', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
-            Command.create({'name': _('Friday Afternoon'), 'dayofweek': '4', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'})
+            Command.create({'name': self.env._('Monday Morning'), 'dayofweek': '0', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
+            Command.create({'name': self.env._('Monday Lunch'), 'dayofweek': '0', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
+            Command.create({'name': self.env._('Monday Afternoon'), 'dayofweek': '0', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
+            Command.create({'name': self.env._('Tuesday Morning'), 'dayofweek': '1', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
+            Command.create({'name': self.env._('Tuesday Lunch'), 'dayofweek': '1', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
+            Command.create({'name': self.env._('Tuesday Afternoon'), 'dayofweek': '1', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
+            Command.create({'name': self.env._('Wednesday Morning'), 'dayofweek': '2', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
+            Command.create({'name': self.env._('Wednesday Lunch'), 'dayofweek': '2', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
+            Command.create({'name': self.env._('Wednesday Afternoon'), 'dayofweek': '2', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
+            Command.create({'name': self.env._('Thursday Morning'), 'dayofweek': '3', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
+            Command.create({'name': self.env._('Thursday Lunch'), 'dayofweek': '3', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
+            Command.create({'name': self.env._('Thursday Afternoon'), 'dayofweek': '3', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
+            Command.create({'name': self.env._('Friday Morning'), 'dayofweek': '4', 'hour_from': 8, 'hour_to': 12, 'day_period': 'morning'}),
+            Command.create({'name': self.env._('Friday Lunch'), 'dayofweek': '4', 'hour_from': 12, 'hour_to': 13, 'day_period': 'lunch'}),
+            Command.create({'name': self.env._('Friday Afternoon'), 'dayofweek': '4', 'hour_from': 13, 'hour_to': 17, 'day_period': 'afternoon'}),
         ]
 
     # --------------------------------------------------
@@ -889,17 +839,16 @@ class ResourceCalendar(models.Model):
                         return revert(start + timedelta(hours=hours))
                     hours -= interval_hours
             return False
-        else:
-            hours = abs(hours)
-            delta = timedelta(days=14)
-            for n in range(100):
-                dt = day_dt - delta * n
-                for start, stop, _meta in reversed(get_intervals(dt - delta, dt)[resource_id]):
-                    interval_hours = (stop - start).total_seconds() / 3600
-                    if hours <= interval_hours:
-                        return revert(stop - timedelta(hours=hours))
-                    hours -= interval_hours
-            return False
+        hours = abs(hours)
+        delta = timedelta(days=14)
+        for n in range(100):
+            dt = day_dt - delta * n
+            for start, stop, _meta in reversed(get_intervals(dt - delta, dt)[resource_id]):
+                interval_hours = (stop - start).total_seconds() / 3600
+                if hours <= interval_hours:
+                    return revert(stop - timedelta(hours=hours))
+                hours -= interval_hours
+        return False
 
     def plan_days(self, days, day_dt, compute_leaves=False, domain=None):
         """
@@ -931,7 +880,7 @@ class ResourceCalendar(models.Model):
                         return revert(stop)
             return False
 
-        elif days < 0:
+        if days < 0:
             days = abs(days)
             found = set()
             delta = timedelta(days=14)
@@ -943,17 +892,7 @@ class ResourceCalendar(models.Model):
                         return revert(start)
             return False
 
-        else:
-            return revert(day_dt)
-
-    def _get_max_number_of_hours(self, start, end):
-        self.ensure_one()
-        if not self.attendance_ids:
-            return 0
-        mapped_data = defaultdict(lambda: 0)
-        for attendance in self.attendance_ids.filtered(lambda a: a.day_period != 'lunch' and ((not a.date_from or not a.date_to) or (a.date_from <= end.date() and a.date_to >= start.date()))):
-            mapped_data[(attendance.week_type, attendance.dayofweek)] += attendance.hour_to - attendance.hour_from
-        return max(mapped_data.values())
+        return revert(day_dt)
 
     def _works_on_date(self, date):
         self.ensure_one()
